README.md
# bits

Converts input into bits (with explanations, where applicable).

## Example

When run with e.g. `bits -0.3`, the output is

```
f16:  1011010011001101
      SEEEEEMMMMMMMMMM
      S: Sign (1 bit)
      E: Exponent (5 bits)
      M: Fraction / Mantissa (10 bits)
bf16: 1011111010011010
      SEEEEEEEEMMMMMMM
      S: Sign (1 bit)
      E: Exponent (8 bits)
      M: Fraction / Mantissa (7 bits)
f32:  10111110100110011001100110011010
      SEEEEEEEEMMMMMMMMMMMMMMMMMMMMMMM
      S: Sign (1 bit)
      E: Exponent (8 bits)
      M: Fraction / Mantissa (23 bits)
f64:  1011111111010011001100110011001100110011001100110011001100110011
      SEEEEEEEEEEEMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
      S: Sign (1 bit)
      E: Exponent (11 bits)
      M: Fraction / Mantissa (52 bits)
```

If only some types are required, the `--type=...` option can be used, e.g. `bin --type=f16,f32 -0.3` which gives

```
f16:  1011010011001101
      SEEEEEMMMMMMMMMM
      S: Sign (1 bit)
      E: Exponent (5 bits)
      M: Fraction / Mantissa (10 bits)
f32:  10111110100110011001100110011010
      SEEEEEEEEMMMMMMMMMMMMMMMMMMMMMMM
      S: Sign (1 bit)
      E: Exponent (8 bits)
      M: Fraction / Mantissa (23 bits)
```

### Full, short and very short format

By default, `--display=full` is implied, showing all information. If only type and bit information are wished,
`--display=short` can be used:

```
f16:  1011010011001101
f32:  10111110100110011001100110011010
```

If only the bits are wished for, the `--display=very-short` format exists, removing even the type information:

```
1011010011001101
10111110100110011001100110011010
```

The very short format makes most sense in combination with specific types; in the example above, `--type=f16,f32`
was used.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2023-11-29

### Added

- Added support for IEEE-754 half precision, single precision and double precision, as well as the brain float format.
- Added console color support.
- Added display modes.

[0.1.0]: https://github.com/sunsided/bits/releases/tag/0.1.0


src/main.rs
use clap::Parser;
#[cfg(feature = "half")]
use half::{bf16, f16};
use std::borrow::Cow;
use std::collections::HashSet;
use std::io::{self, Read, Write};
use std::str::FromStr;
use termcolor::{Color, ColorChoice, ColorSpec, StandardStream, WriteColor};

const PADDING: usize = 6;

#[derive(Debug, clap::Parser)]
#[clap(name = "bits")]
struct Options {
    /// The input to use. Will be ignored if --stdin flag is used.
    #[clap(group = "input_source", allow_hyphen_values = true)]
    input: Option<String>,
    /// If set, read input from stdin instead of positional argument.
    #[clap(long, group = "input_source")]
    stdin: bool,
    /// Set color choice. One of: always, always-ansi, auto, or never.
    #[clap(long, value_parser(color_style))]
    color: Option<ColorChoice>,
    /// Set the allowed type.
    #[clap(long("type"), value_parser(allowed_type), use_value_delimiter(true))]
    types: Option<Vec<AllowedType>>,
    /// Set the allowed type.
    #[clap(long, value_parser(display_mode))]
    display: Option<DisplayMode>,
}

fn main() {
    let options = Options::parse();

    let input = match read_input(options.stdin, &options.input) {
        None => return,
        Some(input) => input,
    };

    let color_choice = options.color.unwrap_or(ColorChoice::Auto);
    let mut stdout = StandardStream::stdout(color_choice);

    let display_mode = options.display.unwrap_or(DisplayMode::Full);

    let types = options.types.map_or(HashSet::new(), HashSet::from_iter);
    let type_enabled = |t: AllowedType| types.is_empty() || types.contains(&t);

    #[cfg(feature = "half")]
    if type_enabled(AllowedType::Ieee754HalfPrecision) {
        try_handle_f16(&input, &mut stdout, display_mode).ok();
    }

    #[cfg(feature = "half")]
    if type_enabled(AllowedType::BrainFloatingPoint) {
        try_handle_bf16(&input, &mut stdout, display_mode).ok();
    }

    if type_enabled(AllowedType::Ieee754SinglePrecision) {
        try_handle_f32(&input, &mut stdout, display_mode).ok();
    }

    if type_enabled(AllowedType::Ieee754DoublePrecision) {
        try_handle_f64(&input, &mut stdout, display_mode).ok();
    }
}

#[cfg(feature = "half")]
fn try_handle_f16<I: AsRef<str>>(
    input: I,
    mut stdout: &mut StandardStream,
    display_mode: DisplayMode,
) -> io::Result<bool> {
    if let Ok(value) = f16::from_str(input.as_ref()) {
        let bits: u16 = unsafe { std::mem::transmute(value) };

        print_colored(
            &mut stdout,
            0,
            format!("f16:  {:016b}", bits),
            [
                if display_mode != DisplayMode::VeryShort {
                    Foreground::Default(6)
                } else {
                    Foreground::Skip(6)
                },
                Foreground::Color(Color::Red, 1),
                Foreground::Color(Color::Green, 5),
                Foreground::Color(Color::Blue, 10),
            ],
        )?;
        if display_mode == DisplayMode::Full {
            print_colored(
                &mut stdout,
                6,
                "SEEEEEMMMMMMMMMM",
                [
                    Foreground::Color(Color::Red, 1),
                    Foreground::Color(Color::Green, 5),
                    Foreground::Color(Color::Blue, 10),
                ],
            )?;
            print_colored(
                &mut stdout,
                6,
                "S: Sign (1 bit)",
                [Foreground::Color(Color::Red, 0)],
            )?;
            print_colored(
                &mut stdout,
                6,
                "E: Exponent (5 bits)",
                [Foreground::Color(Color::Green, 0)],
            )?;
            print_colored(
                &mut stdout,
                6,
                "M: Fraction / Mantissa (10 bits)",
                [Foreground::Color(Color::Blue, 0)],
            )?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

#[cfg(feature = "half")]
fn try_handle_bf16<I: AsRef<str>>(
    input: I,
    mut stdout: &mut StandardStream,
    display_mode: DisplayMode,
) -> io::Result<bool> {
    if let Ok(value) = bf16::from_str(input.as_ref()) {
        let bits: u16 = unsafe { std::mem::transmute(value) };

        print_colored(
            &mut stdout,
            0,
            format!("bf16: {:016b}", bits),
            [
                if display_mode != DisplayMode::VeryShort {
                    Foreground::Default(PADDING)
                } else {
                    Foreground::Skip(PADDING)
                },
                Foreground::Color(Color::Red, 1),
                Foreground::Color(Color::Green, 8),
                Foreground::Color(Color::Blue, 7),
            ],
        )?;
        if display_mode == DisplayMode::Full {
            print_colored(
                &mut stdout,
                PADDING,
                "SEEEEEEEEMMMMMMM",
                [
                    Foreground::Color(Color::Red, 1),
                    Foreground::Color(Color::Green, 8),
                    Foreground::Color(Color::Blue, 7),
                ],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "S: Sign (1 bit)",
                [Foreground::Color(Color::Red, 0)],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "E: Exponent (8 bits)",
                [Foreground::Color(Color::Green, 0)],
            )?;
            print_colored(
                &mut stdout,
                6,
                "M: Fraction / Mantissa (7 bits)",
                [Foreground::Color(Color::Blue, 0)],
            )?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

fn try_handle_f32<I: AsRef<str>>(
    input: I,
    mut stdout: &mut StandardStream,
    display_mode: DisplayMode,
) -> io::Result<bool> {
    if let Ok(value) = f32::from_str(input.as_ref()) {
        let bits: u32 = unsafe { std::mem::transmute(value) };

        print_colored(
            &mut stdout,
            0,
            format!("f32:  {:032b}", bits),
            [
                if display_mode != DisplayMode::VeryShort {
                    Foreground::Default(PADDING)
                } else {
                    Foreground::Skip(PADDING)
                },
                Foreground::Color(Color::Red, 1),
                Foreground::Color(Color::Green, 8),
                Foreground::Color(Color::Blue, 23),
            ],
        )?;
        if display_mode == DisplayMode::Full {
            print_colored(
                &mut stdout,
                PADDING,
                "SEEEEEEEEMMMMMMMMMMMMMMMMMMMMMMM",
                [
                    Foreground::Color(Color::Red, 1),
                    Foreground::Color(Color::Green, 8),
                    Foreground::Color(Color::Blue, 23),
                ],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "S: Sign (1 bit)",
                [Foreground::Color(Color::Red, 0)],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "E: Exponent (8 bits)",
                [Foreground::Color(Color::Green, 0)],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "M: Fraction / Mantissa (23 bits)",
                [Foreground::Color(Color::Blue, 0)],
            )?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

fn try_handle_f64<I: AsRef<str>>(
    input: I,
    mut stdout: &mut StandardStream,
    display_mode: DisplayMode,
) -> io::Result<bool> {
    if let Ok(value) = f64::from_str(input.as_ref()) {
        let bits: u64 = unsafe { std::mem::transmute(value) };

        print_colored(
            &mut stdout,
            0,
            format!("f64:  {:064b}", bits),
            [
                if display_mode != DisplayMode::VeryShort {
                    Foreground::Default(PADDING)
                } else {
                    Foreground::Skip(PADDING)
                },
                Foreground::Color(Color::Red, 1),
                Foreground::Color(Color::Green, 11),
                Foreground::Color(Color::Blue, 52),
            ],
        )?;
        if display_mode == DisplayMode::Full {
            print_colored(
                &mut stdout,
                PADDING,
                "SEEEEEEEEEEEMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM",
                [
                    Foreground::Color(Color::Red, 1),
                    Foreground::Color(Color::Green, 11),
                    Foreground::Color(Color::Blue, 52),
                ],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "S: Sign (1 bit)",
                [Foreground::Color(Color::Red, 0)],
            )
            .ok();
            print_colored(
                &mut stdout,
                PADDING,
                "E: Exponent (11 bits)",
                [Foreground::Color(Color::Green, 0)],
            )?;
            print_colored(
                &mut stdout,
                PADDING,
                "M: Fraction / Mantissa (52 bits)",
                [Foreground::Color(Color::Blue, 0)],
            )?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

fn print_colored<S: AsRef<str>, C: IntoIterator<Item = Foreground>>(
    stdout: &mut StandardStream,
    padding: usize,
    input: S,
    colors: C,
) -> io::Result<()> {
    write!(stdout, "{}", " ".repeat(padding))?;

    let input = input.as_ref();
    let mut start = 0;
    for color in colors {
        let (color, repetitions, show) = match color {
            Foreground::Color(color, repetitions) => (Some(color), repetitions, true),
            Foreground::Default(repetitions) => (None, repetitions, true),
            Foreground::Skip(repetitions) => (None, repetitions, false),
        };

        if !show {
            start += repetitions;
            continue;
        }

        let repetitions = if repetitions == 0 {
            input.len() - start
        } else {
            repetitions
        };
        let end = start + repetitions;
        let slice = &input[start..end];
        start = end;
        stdout.set_color(ColorSpec::new().set_fg(color))?;
        write!(stdout, "{}", slice)?
    }

    stdout.reset()?;

    if start < input.len() {
        let slice = &input[start..];
        write!(stdout, "{}", slice)?
    }

    writeln!(stdout, "")?;
    Ok(())
}

enum Foreground {
    Color(Color, usize),
    Default(usize),
    Skip(usize),
}

fn read_input(stdin: bool, input: &Option<String>) -> Option<Cow<str>> {
    if stdin {
        let mut buffer = String::new();
        io::stdin().read_to_string(&mut buffer).unwrap();
        Some(Cow::Owned(buffer.trim().to_string()))
    } else if let Some(input) = &input {
        Some(Cow::Borrowed(&input))
    } else {
        None
    }
}

fn color_style(s: &str) -> Result<ColorChoice, String> {
    match s {
        "always" => Ok(ColorChoice::Always),
        "always-ansi" => Ok(ColorChoice::AlwaysAnsi),
        "auto" => Ok(ColorChoice::Auto),
        "never" => Ok(ColorChoice::Never),
        _ => Err(String::from(
            "Either never, auto, always or always-ansi must be specified",
        )),
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
enum AllowedType {
    /// Enable [`f16`].
    #[cfg(feature = "half")]
    Ieee754HalfPrecision,
    /// Enable [`bf16`].
    #[cfg(feature = "half")]
    BrainFloatingPoint,
    /// Enable [`f32`].
    Ieee754SinglePrecision,
    /// Enable [`f64`].
    Ieee754DoublePrecision,
}

fn allowed_type(s: &str) -> Result<AllowedType, String> {
    match s {
        #[cfg(feature = "half")]
        "f16" => Ok(AllowedType::Ieee754HalfPrecision),
        #[cfg(feature = "half")]
        "bf16" => Ok(AllowedType::BrainFloatingPoint),
        "f32" => Ok(AllowedType::Ieee754SinglePrecision),
        "f64" => Ok(AllowedType::Ieee754DoublePrecision),
        _ => Err(String::from(
            "Either f16, bf16, f32 or f64 must be specified",
        )),
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
enum DisplayMode {
    /// Shows only bit information (no types).
    VeryShort,
    /// Shows only type and bit information.
    Short,
    /// Shows all information.
    Full,
}

fn display_mode(s: &str) -> Result<DisplayMode, String> {
    match s {
        "very-short" => Ok(DisplayMode::VeryShort),
        "short" => Ok(DisplayMode::Short),
        "full" => Ok(DisplayMode::Full),
        _ => Err(String::from("Either short or full must be specified")),
    }
}
